File: 백준/Gold/7569. 토마토/토마토.js
// 멀티소스
const readline = require("readline").createInterface({
  input: process.stdin,
  output: process.stdout,
});

let input = [];

class Queue {
  constructor() {
    this.items = [];
    this.head = 0;
    this.tail = 0;
  }
  get length() {
    return this.tail - this.head;
  }
  en(item) {
    this.items[this.tail++] = item;
  }
  de() {
    return this.head < this.tail ? this.items[this.head++] : null;
  }
}

const solution = (m, n, h, t) => {
  const queue = new Queue();

  let zCount = 0;
  for (let i = 0; i < h; i++)
    for (let j = 0; j < n; j++)
      for (let k = 0; k < m; k++) {
        if (t[i][j][k] == 0) zCount++;
        if (t[i][j][k] == 1) queue.en([i, j, k, 0]);
      }

  let days = 0;
  const dirs = [
    [1, 0, 0],
    [-1, 0, 0],
    [0, 1, 0],
    [0, -1, 0],
    [0, 0, 1],
    [0, 0, -1],
  ];

  while (queue.length > 0) {
    const [z, y, x, day] = queue.de();
    days = Math.max(days, day);

    for (const [dz, dy, dx] of dirs) {
      const nz = dz + z,
        ny = dy + y,
        nx = dx + x;

      if (
        nz < 0 ||
        ny < 0 ||
        nx < 0 ||
        nz >= h ||
        ny >= n ||
        nx >= m ||
        t[nz][ny][nx] != 0
      )
        continue;

      t[nz][ny][nx] = 1;
      zCount--;
      queue.en([nz, ny, nx, day + 1]);
    }
  }

  return zCount > 0 ? -1 : days;
};

readline
  .on("line", function (line) {
    input.push(line);
  })
  .on("close", function () {
    const [m, n, h] = input[0].split(" ").map((v) => +v);
    const lines = input.slice(1);

    const t = [];
    for (let i = 0; i < lines.length; i++) {
      const floor = Math.floor(i / n);
      t[floor] ??= [];
      t[floor].push(lines[i].split(" ").map((v) => +v));
    }

    console.log(solution(m, n, h, t));
  });

File: 백준/Gold/7569. 토마토/토마토.py
# 다중 확산
import sys
from collections import deque

input = sys.stdin.readline
n, m, h = list(map(int, input().split()))
tomato = []
for _ in range(h):
    floor = []
    for _ in range(m):
        floor.append(list(map(int, input().split())))
    tomato.append(floor)

dirZ = [1, -1, 0, 0, 0, 0]
dirY = [0, 0, 1, -1, 0, 0]
dirX = [0, 0, 0, 0, 1, -1]

q = deque()
for z in range(h):
    for y in range(m):
        for x in range(n):
            if tomato[z][y][x] == 1:
                q.append((z, y, x))

while q:
    cz, cy, cx = q.popleft()
    day = tomato[cz][cy][cx]
    for i in range(6):
        dz, dy, dx = dirZ[i], dirY[i], dirX[i]
        nz, ny, nx = dz + cz, dy + cy, dx + cx
        if nz < 0 or ny < 0 or nx < 0 or nz >= h or ny >= m or nx >= n:
            continue
        if tomato[nz][ny][nx] != 0:
            continue
        tomato[nz][ny][nx] = day + 1
        q.append((nz, ny, nx))


def get_day():
    max_day = 0
    for z in range(h):
        for y in range(m):
            for x in range(n):
                if tomato[z][y][x] == 0:
                    return -1
                max_day = max(max_day, tomato[z][y][x] - 1)
    return max_day


print(get_day())
